go/vt/vttablet/tabletserver/tabletserver.go

```diff
@@ -1372,26 +1365,24 @@ func (tsv *TabletServer) convertAndLogError(ctx context.Context, sql string, bin
 		sqlState := sqlErr.SQLState()
 		errnum := sqlErr.Number()
 		if tsv.TerseErrors && len(bindVariables) != 0 && errCode != vtrpcpb.Code_FAILED_PRECONDITION {
-			err = vterrors.Errorf(errCode, "(errno %d) (sqlstate %s)%s: %s", errnum, sqlState, callerID, queryAsString(sql, nil))
+			err = vterrors.Errorf(errCode, "(errno %d) (sqlstate %s)%s: %s", errnum, sqlState, callerID, queryAsString(sql, bindVariables, tsv.TerseErrors))
 			if logMethod != nil {
-				message = fmt.Sprintf("(errno %d) (sqlstate %s)%s: %s", errnum, sqlState, callerID, truncateSQLAndBindVars(sql, nil))
+				// TerseErrors should generally NOT impact log messages -- rather client errors -- but keeping
+				// this existing exception in place so as not to change the behavior for users of the flag.
+				message = fmt.Sprintf("(errno %d) (sqlstate %s)%s: %s", errnum, sqlState, callerID, truncateSQLAndBindVars(sql, bindVariables, tsv.TerseErrors))
 			}
 		} else {
-			err = vterrors.Errorf(errCode, "%s (errno %d) (sqlstate %s)%s: %s", sqlErr.Message, errnum, sqlState, callerID, queryAsString(sql, bindVariables))
+			err = vterrors.Errorf(errCode, "%s (errno %d) (sqlstate %s)%s: %s", sqlErr.Message, errnum, sqlState, callerID, queryAsString(sql, bindVariables, tsv.TerseErrors))
```

Comment:
I think there is a behavioural change here for TerseErrors. If there is a statement with bindVariables and errCode is `FAILED_PRECONDITION`, the old codepath would not have redacted the bindVariables, but it will now. We should pass `false` to the `queryAsString` function call here.

Reply:
Correct, it will now redact the bind variables if TerseErrors is true. But I would say that this old behavior was a bug, no? The whole point of the flag was to redact bind variables in client error messages, wasn't it?

Reply:
It seems like the assumption made in the previous code as that if we're here, either TerseErrors is OFF *or* we have 0 bind variables so redaction does not matter.

Reply:
After having read that comment... about why the `FAILED_PRECONDITION` handling was in place I agree with you. We need to change this behavior back.

Reply:
This is addressed here: https://github.com/vitessio/vitess/pull/9634

Reply:
I agree 💯 %